exporters/otlp/logs/src/main/java/io/opentelemetry/exporter/otlp/logs/OtlpGrpcLogExporter.java

```diff
@@ -0,0 +1,187 @@
+      Attributes.of(EXPORTER_KEY, EXPORTER_NAME);
+  private static final Attributes EXPORT_SUCCESS_ATTRIBUTES =
+      Attributes.of(EXPORTER_KEY, EXPORTER_NAME, SUCCESS_KEY, "true");
+  private static final Attributes EXPORT_FAILURE_ATTRIBUTES =
+      Attributes.of(EXPORTER_KEY, EXPORTER_NAME, SUCCESS_KEY, "false");
+
+  private static final Logger internalLogger =
+      Logger.getLogger(OtlpGrpcLogExporter.class.getName());
+
+  private final ThrottlingLogger logger = new ThrottlingLogger(internalLogger);
+
+  private final MarshalerLogsServiceGrpc.LogsServiceFutureStub logsService;
+
+  private final ManagedChannel managedChannel;
+  private final long timeoutNanos;
+  private final BoundLongCounter logsSeen;
+  private final BoundLongCounter logsExportedSuccess;
+  private final BoundLongCounter logsExportedFailure;
+
+  /**
+   * Creates a new OTLP gRPC Logs Reporter with the given name, using the given channel.
+   *
+   * @param channel the channel to use when communicating with the OpenTelemetry Collector.
+   * @param timeoutNanos max waiting time for the collector to process each log batch. When set to 0
+   *     or to a negative value, the exporter will wait indefinitely.
+   * @param compressionEnabled whether or not to enable gzip compression.
+   */
+  OtlpGrpcLogExporter(ManagedChannel channel, long timeoutNanos, boolean compressionEnabled) {
+    // TODO: telemetry schema version.
+    Meter meter = GlobalMeterProvider.get().meterBuilder("io.opentelemetry.exporters.otlp").build();
+    this.logsSeen =
+        meter.counterBuilder("logsSeenByExporter").build().bind(EXPORTER_NAME_Attributes);
+    LongCounter logsExportedCounter = meter.counterBuilder("logsExportedByExporter").build();
+    this.logsExportedSuccess = logsExportedCounter.bind(EXPORT_SUCCESS_ATTRIBUTES);
+    this.logsExportedFailure = logsExportedCounter.bind(EXPORT_FAILURE_ATTRIBUTES);
+    this.managedChannel = channel;
+    this.timeoutNanos = timeoutNanos;
+    Codec codec = compressionEnabled ? new Codec.Gzip() : Codec.Identity.NONE;
+    this.logsService =
+        MarshalerLogsServiceGrpc.newFutureStub(channel).withCompression(codec.getMessageEncoding());
+  }
+
+  /**
+   * Submits all the given logs in a single batch to the OpenTelemetry collector.
+   *
+   * @param logs the list of sampled logs to be exported.
+   * @return the result of the operation
+   */
+  @Override
+  public CompletableResultCode export(Collection<LogRecord> logs) {
+    logsSeen.add(logs.size());
+    LogsRequestMarshaler request = LogsRequestMarshaler.create(logs);
+
+    final CompletableResultCode result = new CompletableResultCode();
+
+    MarshalerLogsServiceGrpc.LogsServiceFutureStub exporter;
+    if (timeoutNanos > 0) {
+      exporter = logsService.withDeadlineAfter(timeoutNanos, TimeUnit.NANOSECONDS);
+    } else {
+      exporter = logsService;
+    }
+
+    Futures.addCallback(
+        exporter.export(request),
```

Comment:
You should make sure to detach grpc Context here to make sure you aren't accidentally attaching to something else that was on the same thread.

grpc Context in recent version is much more sensitive to re-use.  Hoping to find time later this week to reproduce a test case for auto-instrumentation.

Reply:
I'm not sure if it applies here. The common use case is the BatchSpanProcessor so the thread is independent and shouldn't affect the app I think. If a user uses the SingleSpanProcessor for some reason (not possible with the javaagent, and generally not recommended by us), I feel like respecting an existing context is actually correct behavior - if we're exporting spans during a request, then that request's timeout actually should be applied to the export request since we're affecting it.

Let me know if this understanding makes sense.